File: serve4j/src/main/java/billow/fun/serve4j/service/StorageService.java
package billow.fun.serve4j.service;

import org.springframework.core.io.Resource;
import org.springframework.core.io.UrlResource;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.net.MalformedURLException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.stream.Stream;

@Service
public class StorageService implements IStorageService {

    private Path root = Path.of("upload-dir");

    @Override
    public void init() {
        if (Files.notExists(root))
            try {
                Files.createDirectories(root);
            } catch (IOException e) {
                System.out.println("failed create $root");
                e.printStackTrace();
            }
    }

    @Override
    public void store(MultipartFile file) {

    }

    @Override
    public Stream<Path> loadAll() {
        try {
            return Files.walk(root, 1)
                    .filter(f -> f != root)
                    .map(f -> root.relativize(f));
        } catch (IOException e) {
            throw new RuntimeException("Failed to read stored files", e);
        }
    }

    @Override
    public Path load(String filename) {
        return root.resolve(filename);
    }

    @Override
    public Resource loadAsResource(String filename) {
        try {
            var file = load(filename);
            var resource = new UrlResource(file.toUri());
            if (resource.exists() || resource.isReadable()) {
                return resource;
            } else {
                throw new RuntimeException("Could not read file: " + filename);
            }
        } catch (MalformedURLException e) {
            throw new RuntimeException("Could not read file: $filename", e);
        }
    }

    @Override
    public void delete(String filename) {
        var path = root.resolve(filename);
        if (Files.exists(path)) {
            try {
                Files.delete(path);
            } catch (IOException e) {
                e.printStackTrace();
            }
        } else {
            throw new RuntimeException("delete failed. $filename is not exist");
        }
    }

    @Override
    public void deleteAll() {
        try {
            Files.createDirectories(root);
        } catch (IOException e) {
            throw new RuntimeException("Could not initialize storage", e);
        }
    }
}

File: server/src/main/kotlin/billow/fun/cms/server/service/StorageService.kt
package billow.`fun`.cms.server.service

import org.springframework.core.io.Resource
import org.springframework.util.StringUtils
import org.springframework.web.multipart.MultipartFile
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.util.stream.Stream
import java.net.MalformedURLException
import org.springframework.core.io.UrlResource
import org.springframework.stereotype.Service


interface StorageService {
    fun init()
    fun store(file: MultipartFile)
    fun loadAll(): Stream<Path>
    fun load(filename: String): Path
    fun loadAsResource(filename: String): Resource
    fun deleteAll()
}

@Service
class FSStorageService : StorageService {
    private val root: Path = Path.of("upload-dir")

    override fun init() {
        if (Files.notExists(root))
            try {
                Files.createDirectories(root)
            } catch (e: IOException) {
                println("failed create $root")
                e.printStackTrace()
            }
    }

    override fun store(file: MultipartFile) {
        val filename = StringUtils.cleanPath(file.originalFilename!!)
        if (file.isEmpty || filename.contains("..")) throw RuntimeException("failed store file" + file.name)
        try {
            val inputStream = file.inputStream
            Files.copy(inputStream, root.resolve(filename), StandardCopyOption.REPLACE_EXISTING)
        } catch (e: IOException) {
            throw RuntimeException("failed store file" + file.name)
        }
    }

    override fun loadAll(): Stream<Path> {
        try {
            return Files.walk(root, 1)
                    .filter { it != root }
                    .map { root.relativize(it) }
        } catch (e: IOException) {
            throw RuntimeException("Failed to read stored files", e)
        }

    }

    override fun load(filename: String): Path = root.resolve(filename)

    override fun loadAsResource(filename: String): Resource {
        try {
            val file = load(filename)
            val resource = UrlResource(file.toUri())
            return if (resource.exists() || resource.isReadable) {
                resource
            } else {
                throw RuntimeException("Could not read file: $filename")
            }
        } catch (e: MalformedURLException) {
            throw RuntimeException("Could not read file: $filename", e)
        }

    }

    override fun deleteAll() {
        try {
            Files.createDirectories(root)
        } catch (e: IOException) {
            throw RuntimeException("Could not initialize storage", e)
        }
    }
}
